src/Agent.Plugins/BuildDrop/BuildDropPlugin.cs

```diff
@@ -0,0 +1,191 @@
+using Agent.Sdk;
+
+namespace Agent.Plugins.BuildDrop
+{
+    public abstract class BuildDropTaskPluginBase : IAgentTaskPlugin
+    {
+        public abstract Guid Id { get; }
+        public string Version => "0.138.0"; // Publish and Download tasks will be always on the same version.
+        public string Stage => "main";
+
+        public async Task RunAsync(AgentTaskPluginExecutionContext context, CancellationToken token)
+        {
+            ArgUtil.NotNull(context, nameof(context));
+
+            // Project ID
+            Guid projectId = new Guid(context.Variables.GetValueOrDefault(BuildVariables.TeamProjectId)?.Value ?? Guid.Empty.ToString());
+            ArgUtil.NotEmpty(projectId, nameof(projectId));
+
+            // Artifact Name
+            string artifactName;
+            if (!context.Inputs.TryGetValue(ArtifactEventProperties.ArtifactName, out artifactName) ||
+                string.IsNullOrEmpty(artifactName))
+            {
+                throw new Exception(StringUtil.Loc("ArtifactNameRequired"));
+            }
+
+            // Path
+            string targetPath;
+            if (!context.Inputs.TryGetValue(ArtifactEventProperties.TargetPath, out targetPath) ||
+                string.IsNullOrEmpty(targetPath))
+            {
+                throw new Exception(StringUtil.Loc("ArtifactLocationRequired"));
+            }
+
+            await ProcessCommandInternalAsync(context, targetPath, projectId, artifactName, token);
+        }
+
+        // Process the command with preprocessed arguments.
+        protected abstract Task ProcessCommandInternalAsync(
+            AgentTaskPluginExecutionContext context, 
+            string targetPath, 
+            Guid projectId,
+            string artifactName, 
+            CancellationToken token);
+    }
+
+    // Caller: PublishBuildDrop task
+    // Can be invoked from a build run or a release run should a build be set as the artifact. 
+    public class PublishBuildDropTask : BuildDropTaskPluginBase
+    {
+        // Same as: https://github.com/Microsoft/vsts-tasks/blob/master/Tasks/PublishBuildDropV0/task.json
+        public override Guid Id => new Guid("ECDC45F6-832D-4AD9-B52B-EE49E94659BE");
+
+        protected override async Task ProcessCommandInternalAsync(
+            AgentTaskPluginExecutionContext context, 
+            string targetPath, 
+            Guid projectId,
+            string artifactName,
+            CancellationToken token)
+        {
+            // Build ID - support publishing in a build run or a release run should the build drop be the artifact.
+            string buildIdStr = context.Variables.GetValueOrDefault(BuildVariables.BuildId)?.Value ?? string.Empty;
+            if (!int.TryParse(buildIdStr, out int buildId))
+            {
+                string hostType = context.Variables.GetValueOrDefault("system.hosttype")?.Value; 
+                if (string.Equals(hostType, "Release", StringComparison.OrdinalIgnoreCase)) {
```

Comment:
there might be a type call DeploymentGroup

Reply:
added